cpp/src/arrow/compute/kernels/scalar_hash_test.cc

```diff
@@ -0,0 +1,336 @@
+      ASSERT_EQ(res1.type()->id(), Type::UINT32);
+    } else {
+      FAIL() << "Unknown function: " << func;
+    }
+    AssertDatumsEqual(res1, res2, true);
+  }
+
+  void CheckPrimitive(const std::string& func, const std::shared_ptr<Array>& arr) {
+    ASSERT_OK_AND_ASSIGN(Datum hash_result, CallFunction(func, {arr}));
+    CheckDeterminisic(func, arr);
+    if (func == "hash64") {
+      AssertHashesEqual<uint64_t>(arr, hash_result, HashPrimitive<uint64_t>(arr));
+    } else if (func == "hash32") {
+      AssertHashesEqual<uint32_t>(arr, hash_result, HashPrimitive<uint32_t>(arr));
+    } else {
+      FAIL() << "Unknown function: " << func;
+    }
+  }
+
+  void CheckBinary(const std::string& func, const std::shared_ptr<Array>& arr) {
+    ASSERT_OK_AND_ASSIGN(Datum hash_result, CallFunction(func, {arr}));
+    CheckDeterminisic(func, arr);
+    if (func == "hash64") {
+      AssertHashesEqual<uint64_t>(arr, hash_result, HashBinaryLike<uint64_t>(arr));
+    } else if (func == "hash32") {
+      AssertHashesEqual<uint32_t>(arr, hash_result, HashBinaryLike<uint32_t>(arr));
+    } else {
+      FAIL() << "Unknown function: " << func;
+    }
+  }
+};
+
+TEST_F(TestScalarHash, Null) {
+  Datum res;
+  std::shared_ptr<Array> arr;
+  std::shared_ptr<Array> exp;
+
+  arr = ArrayFromJSON(null(), R"([])");
+  exp = ArrayFromJSON(uint32(), "[]");
+  ASSERT_OK_AND_ASSIGN(res, CallFunction("hash32", {arr}));
+  AssertArraysEqual(*res.make_array(), *exp);
+  CheckDeterminisic("hash32", arr);
+
+  arr = ArrayFromJSON(null(), R"([])");
+  exp = ArrayFromJSON(uint64(), "[]");
+  ASSERT_OK_AND_ASSIGN(res, CallFunction("hash64", {arr}));
+  AssertArraysEqual(*res.make_array(), *exp);
+  CheckDeterminisic("hash64", arr);
+
+  arr = ArrayFromJSON(null(), R"([null, null, null])");
+  exp = ArrayFromJSON(uint32(), "[0, 0, 0]");
+  ASSERT_OK_AND_ASSIGN(res, CallFunction("hash32", {arr}));
+  AssertArraysEqual(*res.make_array(), *exp);
+  CheckDeterminisic("hash32", arr);
+
+  arr = ArrayFromJSON(null(), R"([null, null, null])");
+  exp = ArrayFromJSON(uint64(), "[0, 0, 0]");
+  ASSERT_OK_AND_ASSIGN(res, CallFunction("hash64", {arr}));
+  AssertArraysEqual(*res.make_array(), *exp);
+  CheckDeterminisic("hash64", arr);
+}
+
+TEST_F(TestScalarHash, NullHashIsZero) {
+  auto arr1 = ArrayFromJSON(int32(), R"([null, 0, 1])");
+  ASSERT_OK_AND_ASSIGN(auto res1, CallFunction("hash64", {arr1}));
+  auto buf1 = res1.array()->GetValues<uint64_t>(1);
+  ASSERT_EQ(buf1[0], 0);
+  ASSERT_NE(buf1[1], 0);
+  ASSERT_NE(buf1[2], 0);
+  ASSERT_NE(buf1[1], buf1[2]);
+
+  auto arr2 = ArrayFromJSON(int8(), R"([null, 0, 1])");
+  ASSERT_OK_AND_ASSIGN(auto res2, CallFunction("hash32", {arr2}));
+  auto buf2 = res2.array()->GetValues<uint32_t>(1);
+  ASSERT_EQ(buf2[0], 0);
+  ASSERT_NE(buf2[1], 0);
+  ASSERT_NE(buf2[2], 0);
+  ASSERT_NE(buf2[1], buf2[2]);
+}
+
+TEST_F(TestScalarHash, Boolean) {
+  auto arr = ArrayFromJSON(boolean(), R"([true, false, null, true, null])");
```

Comment:
Should also check that "true", "false" and "null" all hash to different values.

Reply:
Now checking all of those.